### PR TITLE
Add sorting bookmarks and tracks by name

### LINKED ISSUE #1066: Sort bookmarks by name
Currently, sorting by name (ascending and descending) is not available.

### PR DESCRIPTION
**Pull request by sklueber**
This PR adds the ability to sort bookmarks and tracks alphabetically by name (A->Z). This also allows a crude way of custom sorting by prefixing names accordingly.

 Tracks and bookmarks are kept as separate blocks. As there is no concept of reversing the sort order yet, I could add the reverse order as an additional sort type but I wanted to leave that up to discussion first.

| Menu item | Result | Result with tracks and bookmarks |
|--------|--------|--------|
|Screenshot_20240115-132758| Screenshot_20240115-132804 | Screenshot_20240115-154930

Fixes (partially) #1066 (as the lists themselves are not sorted) and partially #4250

### CODE REVIEW

File: data/strings/strings.txt [hunk 1]
```diff
@@ -24991,6 +24991,12 @@
     zh-Hans = 按时间
     zh-Hant = 按時間
 
+  [by_name]
+    comment = Android
+    tags = android
+    en = By name
```
Comment (biodranik) [inline data/strings/strings.txt:24997]:
```
    en = By Name
    af = Op Naam
    ar = بالاسم
    az = Ada görə
    be = Па імені
    bg = По име
    ca = Per nom
    cs = Podle jména
    da = efter navn
    de = Nach Name
    el = Με όνομα
    es = Por nombre
    es-MX = Por nombre
    et = Nime järgi
    eu = Izenaren arabera
    fa = بر اساس اسم
    fi = Nimen mukaan
    fr = Par nom
    he = לפי שם
    hi = नाम द्वारा
    hu = Név szerint
    id = Berdasarkan Nama
    it = Per nome
    ja = 名前で
    ko = 이름별
    lt = Pagal pavadinimą
    mr = नावाने
    nb = Etter navn
    nl = Op naam
    pl = Według nazwy
    pt = Por nome
    ro = După nume
    ru = По имени
    sk = Podľa mena
    sv = Efter namn
    sw = Kwa Jina
    th = ตามชื่อ
    tr = İsme göre
    uk = За назвою
    vi = Theo Tên
    zh-Hans = 按名称
    zh-Hant = 按名字
```

File: android/app/src/main/java/app/organicmaps/bookmarks/ChooseBookmarksSortingTypeFragment.java [hunk 2]
```diff
@@ -88,7 +90,11 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
     if (args == null)
       throw new AssertionError("Arguments of choose sorting type view can't be null.");
 
-    UiUtils.hide(view, R.id.sort_by_type, R.id.sort_by_distance, R.id.sort_by_time);
+    UiUtils.hide(view,
```
Comment (biodranik) [inline android/app/src/main/java/app/organicmaps/bookmarks/ChooseBookmarksSortingTypeFragment.java:93]: 120 chars in line is ok.

**Review by biodranik — COMMENTED** [review]
Thanks!

1. What are the issues with the implementation?
2. What about the performance?
3. Is it feasible to sort entries in memory once, save them in the correct order into the KML file (internally, bookmarks are stored in KML), and then load/restore the sorted order from that file directly?

**Comment by sklueber** [comment]
> 1. What are the issues with the implementation?

To be quite honest, I just copied the structure of existing sort types. It's essentially the same as `sortByTime`, just with simplifications because I don't divide into subcategories like it does.

So any fundamental issues would be shared among all of the existing sorting.

It uses the "preferred name" that is also used to display the bookmark so everything should be sorted as expected.


> 2. What about the performance?

The 690 bookmark file from the test data folder appears to be sorted instantly as far as I can tell on my Pixel 6a (as opposed to the distance sort which does take a second, since it has to calculate distances first).

> 3. Is it feasible to sort entries in memory once, save them in the correct order into the KML file (internally, bookmarks are stored in KML), and then load/restore the sorted order from that file directly?

That would delete the "default order", wouldn't it? Either way, users can rename or add bookmarks at any time and the sort has to reflect that so I don't believe introducing I/O would help.

Again, assuming the existing sort implementations are acceptable, this shouldn't be any different.

File: data/strings/strings.txt (hunk 1 above)
Comment (sklueber) [reply data/strings/strings.txt:24997]: Having "by **N**ame" capitalized is inconsistent with the other sorting options, so I kept it lowercase

File: map/bookmark_manager.cpp [hunk 3]
```diff
@@ -1231,6 +1251,35 @@ void BookmarkManager::SortByType(std::vector<SortBookmarkData> const & bookmarks
...
+                                 SortedBlocksCollection & sortedBlocks)
+{
+  std::vector<SortTrackData> sortedTracks = tracksForSort;
+  SortTracksByName(sortedTracks);
+  AddTracksSortedBlock(sortedTracks, sortedBlocks);
+
+  std::vector<SortBookmarkData const *> sortedMarks;
+  sortedMarks.reserve(bookmarksForSort.size());
+  for (auto const & mark : bookmarksForSort)
+      sortedMarks.push_back(&mark);
```
Comment (biodranik) [inline map/bookmark_manager.cpp:1265]:
```suggestion
    sortedMarks.push_back(&mark);
```

File: map/bookmark_manager.cpp [hunk 4]
```diff
@@ -1231,6 +1251,35 @@ void BookmarkManager::SortByType(std::vector<SortBookmarkData> const & bookmarks
...
+              return lbm->m_name < rbm->m_name;
+            });
+
+  // Put all bookmarks into one block
+  SortedBlock bookmarkBlock;
+  bookmarkBlock.m_blockName = GetBookmarksSortedBlockName();
+  for (auto mark : sortedMarks)
+  {
+      bookmarkBlock.m_markIds.push_back(mark->m_id);
+  }
```
Comment (biodranik) [inline map/bookmark_manager.cpp:1279]:
```suggestion
    bookmarkBlock.m_markIds.push_back(mark->m_id);
```

File: map/bookmark_manager.cpp [hunk 5]
```diff
@@ -544,7 +547,8 @@ std::vector<BookmarkManager::SortingType> BookmarkManager::GetAvailableSortingTy
     sortingTypes.push_back(SortingType::ByDistance);
   if (byTimeChecked)
     sortingTypes.push_back(SortingType::ByTime);
-
+  // sorting by name should always be possible
```
Comment (biodranik) [inline map/bookmark_manager.cpp:550]: nit:
```suggestion
  // Sorting by name should always be possible.
```

**Review by biodranik — COMMENTED** [review]
@sklueber I'm very sorry for the delay with the review. Let's fix and merge this important improvement!

Please check and fix failing checks, for example https://github.com/organicmaps/organicmaps/actions/runs/7547039543/job/20805713595#step:6:6262

**Comment by biodranik** [comment]
Closing this as the PR was already merged. TODOs are tracked here: https://github.com/organicmaps/organicmaps/issues/1066#issuecomment-2037413621

**Pull request closed without merging**